Automated deployment

lentinj opened the pull request:
Upgrade python/npm, automate much more of the deployment

@hyanwong https://github.com/OneZoom/OZtree/commit/045217c9e68a625603d486c4b87e84cce551c1ba it'd be useful to see what you think about. I'm presuming that is_testing iff rocket is a good assumption, but may not always be true under beta, e.g. It'd be easy to add a config override if that's the case.

lentinj committed:
Gruntfile: Switch to python3.11

lentinj committed:
package.json: Support Node 18.x

To run webpack on node 18 we need to set --openssl-legacy-provider
until we can upgrade webpack.

lentinj committed:
Gruntfile: web2py_configure for web2py housekeeping

Automate housekeeping for web2py in a Grunt task

lentinj committed:
Gruntfile: venv for web2py #676

We need an up-do-date pymysql, get it by wrapping a venv around web2py

Add a web2py-run helper to use when running standalone scripts, that
will get the web2py setup right.

lentinj committed:
controllers/API: Remove search_log

This will impact search performance, by forcing all search queries to
go through the unique index before searching.

Remove, replace with google analytics later.

lentinj committed:
models/db: is_testing iff running under rocket

Instead of having to tweak is_testing for production use, check the
request environment to see what server we're using.

lentinj committed:
models/db: Database session-handling #82

Not that we should be using sessions much, if at all.

lentinj committed:
install: Install scripts for nginx/supervisord

Scripts to configure nginx/supervisord

hyanwong commented:
Yeah, we set is_testing=True on beta, which can be useful at times.

We can't set it in appconfig.ini, because one of the major speedups from `is_testing=False` is that it only reads appconfig.ini once, at the start of firing up web2py.

lentinj commented:
> Yeah, we set is_testing=True on beta, which can be useful at times.

Ideally IMO beta would be "as-live" as possible, just with a different server name. Setting ``is_testing`` here doesn't feel ideal.

But the above doesn't stop this, rather it'd be the beta server that needs to code altered, not production, which feels the right way round.

> We can't set it in appconfig.ini, because one of the major speedups from is_testing=False is that it only reads appconfig.ini once, at the start of firing up web2py.

I was thinking of reloading the config if is_testing is set to get around this, but enabling/disabling would be a bit weird in retrospect (enabling would require a restart, disabling wouldn't).

Another option would be hooking into ``grunt dev`` vs ``grunt prod`` somehow, but I suspect that's getting carried away.

lentinj committed:
Gruntfile: Alter web2py.py to use venv #676

Update shebang at the top of web2py.py so it uses the venv by default.

davidebbo commented:
I don't know much about how these production branches are used, but why is this new one based on the old one that has a last commit of September 2022? Shouldn't it be rebased on the latest main?

lentinj commented:
Hrm, good point.

We're in the progress of migrating onezoom.org, so I want to deploy the current live site to the new host, not an updated version based on [``main``](https://github.com/OneZoom/OZtree/tree/main). The theory goes that the [``production``](https://github.com/OneZoom/OZtree/tree/production) branch is the state of the live server, which is why this is trying to merge to [``production``](https://github.com/OneZoom/OZtree/tree/production).

Turns out it's not. Assuming I'm looking in the right place it seems to be a local commit off https://github.com/OneZoom/OZtree/commit/d70c6896. @hyanwong any reason not to update the production branch?

hyanwong commented:
Basically, I wanted all instances to share the same set of thumbnail image directories. These would not be set up using automatic deployment, as collecting all the images would be impossible during deployment (many have, in fact, vanished from the Internet now).

lentinj commented:
> > We need to handle images.onezoom.org, although currently live/beta don't use it, maybe they should?
> 
> The images.onezoom.org thing has been a useful redirect in the past. I've tried to keep stuff in the same filesystem. That means `images.onezoom.org` can point to a specific directory, and then both Beta and Prod can symlink to that directory, so they are all using the same thumbnails. I guess this would be tricky if these are all in their own jail.

True, the ``static/FinalOutputs/img`` symlinks wouldn't work across jails, but in development (and in the example config) we set ``url_base = //images.onezoom.org/``. If this was set on beta & prod there'd be no reason for the symlinks, both sites would use the same image URLs from the get-go.

If we wanted to stop image-sharing, setting ``url_base = //imagesbeta.onezoom.org/`` or disabling would do the trick.

There's potentially cross-origin problems to deal with, but if there were then we'd have noticed long ago when developing.

Another nice feature of ``images.onezoom.org`` is it allows us to have alternate hosting for the images, in the future, throwing them through cloudflare or summat.

lentinj committed:
README.markdown: Rework installation instructions

Remove dependencies that are now part of tree-build, rework
instructions to take into account new Grunt rules.

Rework based on the

lentinj commented:
@hyanwong I've had a bash at rewriting the README a bit to reflect currently reality. Mind having a read through and see if it makes sense? It's probably easier to look at https://github.com/OneZoom/OZtree/tree/production-next than the above commit.

davidebbo commented on README.markdown:
```diff
@@ -10,220 +10,134 @@ There are two ways in which you can install OneZoom on a personal computer: full
 
 * *Full installation* creates an entire duplicate of the OneZoom website, which is built using the [web2py](http://web2py.com) framework. This creates a fully self-contained local system (apart from the picture files, which can be downloaded separately). This is the most reliable installation method, but requires you to install and run extra software packages, in particular [web2py](http://web2py ...
 
-
```
Can't comment on existing lines, but in first paragraph it links to `README_SERVER.markdown`, which doesn't seem to exist.

davidebbo commented on README.markdown:
```diff
@@ -10,220 +10,134 @@ There are two ways in which you can install OneZoom on a personal computer: full
-	* create indexes on the tables by running the SQL script in `OZtree/OZprivate/ServerScripts/SQL/create_db_indexes.sql`. You can do this, for example, by running `SOURCE /path/to/OZtree/OZprivate/ServerScripts/SQL/create_db_indexes.sql` within a mysql client.
-	* set `is_testing = False` in `models/db.py` and `migrate=0` in appconfig.ini.
 
+For any installation, OneZoom requires node & python (3.11).
```
I thought we had settled on 3.10 for now.

lentinj replied:
Hrm [I took @hyanwong's last comment here](https://github.com/OneZoom/OZtree/issues/72#issuecomment-1864335699) to imply 3.11 was the way?

Is there another thread discussing this?

lentinj replied:
Yes there is, https://github.com/OneZoom/OZtree/pull/699. But there doesn't seem to be much rationale of 3.10 vs 3.11. Anyways, it's not a big deal.

davidebbo commented on README.markdown:
```diff
@@ -10,220 +10,134 @@ There are two ways in which you can install OneZoom on a personal computer: full
+### Full installation
 
-## Building the OneZoom tree viewer
+In addition, for a full installation you also need nginx, supervisor & MySQL.
```
Maybe the word 'full' is overloaded. To run a 'full' server per the definition above, you don't need nginx. But you'll need to to run a **production** server. Probably the same goes for supervisor.

davidebbo commented on README.markdown:
```diff
@@ -10,220 +10,134 @@ There are two ways in which you can install OneZoom on a personal computer: full
 
-#### Compile documentation
-`grunt docs`: Use this command to generate a compiled documentation file. This will generate a large compiled markdown file in `OZprivate/rawJS/OZTreeModule/docs/_compiled.markdown`, which is best viewed once you have got web2py running, by pointing your browser to `dev/DOCS` (e.g. at `http://127.0.0.1:8000/dev/DOCS`).  Note that viewing this page requires a working internet connection to get var ...
+On Windows we recommended downloading the MSI installer as it will make it easier to configure the new server during the installation
```
Clarify that this refers to the MySQL MSI, and ideally link to it.

Also, missing period after installation.

davidebbo commented on README.markdown:
```diff
@@ -10,220 +10,134 @@ There are two ways in which you can install OneZoom on a personal computer: full
+
+## Database setup / migation
+
+For full installations, you need to setup your database.
```
nit: 'set up' :)

davidebbo commented on README.markdown:
```diff
@@ -10,220 +10,134 @@ There are two ways in which you can install OneZoom on a personal computer: full
+## Database explorer
 
-On the OneZoom main site, web2py is run using a combination of nginx and uwsgi. This is complete overkill if you just want to run a local copy of OneZoom for testing purposes. You can simply run a [temporary and basic web2py server using Python 3](http://www.web2py.com/books/default/chapter/29/03/overview#Startup). The simplest is to open a command-line prompt in the root web2py folder, and run t ...
+(optional) We find it useful to have a GUI interface to connect to the database and run SQL scripts, this can be used instead of using MySQL command line (similar to Windows command line) that is installed by default with MySQL. On Mac OS X we use the (excellent) http://www.sequelpro.com. On windows you could try http://www.mysql.com/products/workbench/ or https://www.quest.com/products/toad-for- ...
```
Worth noting that when using WSL2, installing SQL Workbench on Windows works great to connect to the Ubuntu MySQL instance.

davidebbo reviewed (commented):
Misc comments, some quite minor.

lentinj committed:
Gruntfile: Use python3.10, not python3.11 #672 #699

lentinj committed:
README: Don't mention temporary branch

Ideally this would be separated out into instructions for a production
instance, but at least not mentioning a temporary branch is good.

lentinj committed:
install-nginx: /var/db/acme/live/ isn't created OOTB

An empty installation doesn't have /var/db/acme/live/, only
/var/db/acme/.

lentinj committed:
install-nginx: BSD doesn't know about ln -r

lentinj committed:
install-nginx: Use NGINX_PATH when generating include config

lentinj committed:
install-nginx: Remove v6 listeners

Nginx falls over if v6 isn't available, which it isn't.

lentinj committed:
install-supervisord: BSD-compatible stat syntax

lentinj committed:
install-supervisord: Correct supervisord conf location on BSD

There is no /usr/local/etc/supervisord, it's all in /usr/local/etc.

lentinj committed:
install-supervisord: Put script in debug mode

It's comforting to see what it's doing, and makes error messages more
intelligable.

lentinj committed:
README.markdown: Separate "production installation" section

Move the production installation notes into their own section, mention
the install scripts to be run as root.

davidebbo commented:
> More seriously, ideally (2a) doesn't exist because tree-build is easy enough there's no reason to do anything else. Could we go as far as tree-build having a Github action generating artefacts that OZtree consumes?

/cc @hyanwong. That's an interesting idea. One challenge is that tree-build requires some massive downloads that take hours (in particular a wikidata dump). That would make the build extremely heavy, unless the files are somehow cached on the build machine. Another angle is that some users may be using their own trees, in which case they'll inevitable need to build their own artifacts in tree-build and consume them in OZtree.
